diff -Nur icu-orig/source/layoutex/ParagraphLayout.cpp icu/source/layoutex/ParagraphLayout.cpp
--- icu-orig/source/layoutex/ParagraphLayout.cpp	2007-12-12 19:58:04.000000000 +0100
+++ icu/source/layoutex/ParagraphLayout.cpp	2009-08-31 12:21:04.000000000 +0200
@@ -813,7 +813,7 @@
 
     return nullLanguageCode;
 }
-#elif
+#else
 
 // TODO - dummy implementation for right now...
 le_int32 ParagraphLayout::getLanguageCode(const Locale *locale)
